Resolves #1307 - BucketListIsConsistentWithDatabase Invariant

jonjove opened the pull request:
Introduces BucketListIsConsistentWithDatabase invariant which resolves #1307. This invariant implements the newly introduced checkOnBucketApply to ensure that the database matches the bucket, after it was applied, for all ledgers starting from the first ledger considered in the bucket. Merge after #1317

MonsieurNicolas commented:
Commit names must be useful when running `git log` (only thing that can reference github issues etc is the PR itself)

MonsieurNicolas commented on src/bucket/Bucket.h:
```diff
@@ -42,7 +43,29 @@ class Bucket : public std::enable_shared_from_this<Bucket>,
     // Helper class that reads through the entries in a bucket, used internally
     // during merging.
-    class InputIterator;
+    class InputIterator
```
couple things:
* in general it's better to have a separate commit for code moves like this ("moved Bucket::InputIterator definition from cpp to h file")
* in this case, I don't see the reason to move it to the .h file, it just pollutes the .h file (it's only used internally), if all you wanted was to stop having inline declarations, you can still achieve the same thing by having both the class definition and implementation in the same file

MonsieurNicolas replied:
nm, I didn't see the rest of the changes in the file - was confused by the comment that is no longer relevant.

MonsieurNicolas replied:
now that you are making this available outside of this file, it's the occasion to remove the code in `BucketApplicator` that does the same thing from what I can tell

MonsieurNicolas commented on src/bucket/Bucket.h:
```diff
@@ -42,7 +43,29 @@ class Bucket : public std::enable_shared_from_this<Bucket>,
     // Helper class that reads through the entries in a bucket, used internally
     // during merging.
-    class InputIterator;
+    class InputIterator
```
I would remove this part of the comment:
> , used internally during merging.

MonsieurNicolas commented on src/invariant/BucketListIsConsistentWithDatabase.cpp:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    uint64_t nAccounts = 0, nTrustLines = 0, nOffers = 0, nData = 0;
+    uint32 modifiedSinceLedgerSeq = std::numeric_limits<uint32>::max();
```
from what I can tell the logic around `modifiedSinceLedgerSeq` is not correct, the value of `modifiedSinceLedgerSeq` is defined by which bucket we're checking in the [context of a bucketlist](https://github.com/stellar/stellar-core/blob/master/src/bucket/BucketList.h#L106-L108)
With the expected range figured out, you can then:
* ensure that the entries in this bucket are in the expected range (this probably belongs in the regular apply code, not in an invariant)
* query the SQL database for entries that fall into the range (well, with the upper bound set to max int this code)

MonsieurNicolas commented on src/invariant/BucketListIsConsistentWithDatabase.cpp:
```diff
@@ -0,0 +1,112 @@
+    uint32 modifiedSinceLedgerSeq = std::numeric_limits<uint32>::max();
+    for ( ; iter; ++iter)
+    {
+        auto& e = *iter;
```
should be `auto const&`

MonsieurNicolas commented on src/invariant/BucketListIsConsistentWithDatabase.h:
```diff
@@ -0,0 +1,33 @@
+class Database;
+class LedgerDelta;
+
+class BucketListIsConsistentWithDatabase : public Invariant
```
Please add some comments to explain what this invariant does

MonsieurNicolas commented on src/invariant/BucketListIsConsistentWithDatabase.cpp:
```diff
@@ -0,0 +1,112 @@
+            case DATA:
+                ++nData;
+                break;
+            }
```
we typically have to add (to avoid warnings)
```
default:
     abort(false);
```

MonsieurNicolas commented:
forgot something else: you need to add a few test cases

MonsieurNicolas commented:
also you need to add an index to the tables that you are querying with `lastmodified`

jonjove commented:
I have updated the branch for people to review but it is not currently ready for merge

MonsieurNicolas commented on src/bucket/BucketTests.cpp:
```diff
@@ -835,6 +835,64 @@ TEST_CASE("bucket persistence over app restart", "[bucket][bucketpersist]")
     }
 }
 
+TEST_CASE("bucket count", "[bucket][count]")
```
I wonder if we should make `kNumLevels` a config (we can still hardcode that the app is never started with something other than 10 outside of the `--test` code path in main), such that we can test the behavior up to kNumLevels. Right now the test cases cannot spill up to 10, and therefore we cannot observe the "steady state" behavior from test cases.

jonjove replied:
In order to create a proper test case for issue #1300, we will need to be able to get entries into the bottom level of the BucketList since it is the only place where a merge can cause the elimination of a DEADENTRY and corresponding LIVEENTRY.

MonsieurNicolas commented on src/bucket/BucketTests.cpp:
```diff
@@ -835,6 +835,64 @@ TEST_CASE("bucket persistence over app restart", "[bucket][bucketpersist]")
+            uint32_t currMin = std::numeric_limits<uint32_t>::max();
+            for (Bucket::InputIterator iter(curr); iter; ++iter)
+            {
+                auto lastModified = EntryFrame::FromXDR((*iter).liveEntry())->getLastModified();
```
you don't need the `FromXDR` here

MonsieurNicolas replied:
(you can access the `lastModified` field directly)

jonjove replied:
Fixed

MonsieurNicolas commented on src/invariant/InvariantTests.cpp:
```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
+            size_t count = 0;
+            {
+                auto prep = db.getPreparedStatement(
+                    "SELECT"
```
I would rather not have raw SQL code in test cases like this, and instead have those statements in the respective class, for example, here you can use the `*Frame::countObjects` methods - in order to avoid duplication, maybe we just need to have those methods take a min and max, and this test can just have `min==max` ?

MonsieurNicolas commented on src/invariant/InvariantTests.cpp:
```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
+        std::string account;
+        {
+            auto prep = db.getPreparedStatement(
+                "SELECT accountid FROM accounts"
```
you don't actually need this, standard load + check that last modified is what you expect does the same without introducing new SQL

MonsieurNicolas replied:
what I am advocating for is to avoid having SQL or soci code outside of the few classes that deal with it - it's fine to have helper methods exposed for tests only

MonsieurNicolas commented on src/invariant/InvariantTests.cpp:
```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
+        uint32_t assetType;
+        {
+            auto prep = db.getPreparedStatement(
+                "SELECT accountid, issuer, assetcode, assettype FROM trustlines"
```
same thing here - use regular code to load from database

MonsieurNicolas commented on src/invariant/InvariantTests.cpp:
```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
+            assert(!st.fetch());
+        }
+
+        LedgerEntry entry;
```
this code goes away if you use the regular load method, right?

jonjove replied:
The same comments as above apply to replacing the above SQL query with the regular load method. If I had a key for an entry with the appropriate ledger number (either through the above SQL query or through maintaining that information internally) and used the key to invoke the regular load method, this code would be simplified significantly. It is not so easy unfortunately. TrustFrame::loadTrustLine takes an Asset as a parameter. But to construct the object of type Asset (defined in XDR), would require much of this exact code specifically the if (assetType == ...) part. In net, little can be gained in terms of removing this code by using the regular load functions.

jonjove commented on src/bucket/BucketList.cpp:
```diff
@@ -171,20 +171,108 @@ BucketList::mask(uint32_t v, uint32_t m)
+
+    auto const size = levelSize(level);
+    auto const half = levelHalf(level);
+    if (mask(ledger, half) != 0 && level != 10)
```
Fixed

jonjove commented on src/invariant/BucketListIsConsistentWithDatabase.cpp:
```diff
@@ -38,21 +39,16 @@ BucketListIsConsistentWithDatabase::getName() const
         auto const& e = *iter;
         if (e.type() == LIVEENTRY)
         {
+            assert(EntryFrame::FromXDR(e.liveEntry())->getLastModified() >= oldestLedger);
```
1. Fixed
2. Good catch. This was originally written as an assert to facilitate debugging, but it really should be a part of the invariant. I can also extend this check to make sure the lastModifiedLedgerSeq is no bigger than the upper bound, if you think that is worthwhile?

jonjove commented on src/historywork/ApplyBucketsWork.cpp:
```diff
@@ -22,43 +22,35 @@ namespace stellar
     // held in firstVerified's state.
-    auto lcl = app.getLedgerManager().getLastClosedLedgerHeader();
-    if (firstVerified.header.ledgerSeq < lcl.header.ledgerSeq)
+    /*auto lcl = app.getLedgerManager().getLastClosedLedgerHeader();
```
Fixed. In my opinion, this check should actually be moved upstream to where ApplyBucketsWork is created in CatchupMinimal. Thoughts?

MonsieurNicolas replied:
yes that sounds good - @vogel will probably want to incorporate this in his changeset when he rebases

vogel replied:
Good idea. We can also always leave assert() here and precondition comment to be extra safe.

jonjove commented on src/historywork/CatchupMinimalWork.cpp:
```diff
@@ -131,10 +131,10 @@ CatchupMinimalWork::onSuccess()
-        CLOG(INFO, "History") << "Catchup MINIMAL applying buckets for state "
-                              << LedgerManager::ledgerAbbrev(mFirstVerified);
+        //CLOG(INFO, "History") << "Catchup MINIMAL applying buckets for state "
+        //                      << LedgerManager::ledgerAbbrev(mFirstVerified);
```
This logging should not be removed, since it includes an important state update for CatchupMinimal. Will uncomment.

MonsieurNicolas commented on src/invariant/InvariantTests.cpp:
```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
+                                                  uint32_t)>& f)
+{
+    auto seed = std::chrono::system_clock::now().time_since_epoch().count();
+    std::default_random_engine gen(seed);
```
forgot to mention: do not mess with the seed in test cases to keep test runs deterministic. If you need more randomness, iterate more.

vogel commented on src/bucket/BucketTests.cpp:
```diff
@@ -835,6 +835,59 @@ TEST_CASE("bucket persistence over app restart", "[bucket][bucketpersist]")
     }
 }
 
+TEST_CASE("bucket count", "[bucket][count]")
```
You are testing sizeOfCurr new methods of BucketList.cpp here, what do you think about adding tests just for them? Make a bunch of tuples {ledger, level, expectedSizeOfCurr, expectedSizeOfSnap, expectedOldestLedgerInCurr, expectedOldestLedgerInSnap} and so on?

jonjove replied:
I can definitely try to do this, but those functions are quite complicated and defined recursively so I would have to figure out how to evaluate them without using the implementations. Perhaps an easier idea is to try to write a test case which validates some invariants among those formulae, such as `oldestLedgerInSnap(ledger, level) + sizeOfSnap(ledger, level) == oldestLedgerInCurr(ledger, level)`. I could also add tests like `sizeOfSnap(ledger, level) == levelHalf(level)` implies `sizeOfSnap(ledger2, level) == levelHalf(level)` for `ledger2 > ledger`. Let me know if you think these kinds of tests are sufficient or if you would prefer me try to do some direct calculations.

vogel replied:
What about invariants and some edge cases?

vogel commented on src/bucket/BucketTests.cpp:
```diff
@@ -835,6 +835,59 @@ TEST_CASE("bucket persistence over app restart", "[bucket][bucketpersist]")
+    BucketList& bl = app->getBucketManager().getBucketList();
+    std::vector<LedgerKey> emptySet;
+
+    for (uint32_t nLedgers = 2; nLedgers <= 256; ++nLedgers)
```
isn't that more like currentLedger instead of "number of ledgers"?

vogel commented on src/bucket/BucketTests.cpp:
```diff
@@ -835,6 +835,59 @@ TEST_CASE("bucket persistence over app restart", "[bucket][bucketpersist]")
+                REQUIRE(currMin == BucketList::oldestLedgerInCurr(nLedgers, level));
+            }
+
+            CLOG(INFO, "Bucket") << "Checking Snap[" << level << "]";
```
looks like exact copy of code above - candidate for function/lamda?

vogel commented on src/historywork/ApplyBucketsWork.cpp:
```diff
@@ -20,51 +22,43 @@ namespace stellar
-    std::shared_ptr<Bucket> b;
-    if (hash.find_first_not_of('0') == std::string::npos)
+    std::shared_ptr<Bucket const> b;
+    if (isZero(hexToBin256(hash)))
```
I like that!

vogel commented on src/invariant/BucketListIsConsistentWithDatabase.cpp:
```diff
@@ -0,0 +1,104 @@
+        uint32_t oldestLedger)
+{
+    uint64_t nAccounts = 0, nTrustLines = 0, nOffers = 0, nData = 0;
+    for (Bucket::InputIterator iter(bucket); iter; ++iter)
```
I think we could add support for for-range for Bucket::InputIterator, should not be hard, but is it worth the time?

vogel commented on src/invariant/BucketListIsConsistentWithDatabase.cpp:
```diff
@@ -0,0 +1,104 @@
+    for (Bucket::InputIterator iter(bucket); iter; ++iter)
+    {
+        auto const& e = *iter;
+        if (e.type() == LIVEENTRY)
```
why are you not testing DEADENTRY here? is that impossible (I'm not very familiar with Buckets code, just curious)

vogel commented on src/invariant/InvariantTests.cpp:
```diff
@@ -122,3 +621,82 @@ TEST_CASE("onLedgerClose fail/succeed", "[invariant]")
     }
 }
+
+TEST_CASE("onBucketApply fail/succeed", "[invariant]")
```
I would prefer to have these tests in separate file - BucketListIsConsistentWithDatabaseTests

vogel commented on src/invariant/InvariantTests.cpp:
```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
+    }
+};
+
+class ApplyBucketsWorkDeleteEntry : public ApplyBucketsWork
```
I think you could also consider making one class that accepts lambda executed once on onSuccess here instead of 3 classes.

vogel commented on src/invariant/InvariantTests.cpp:
```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
+
+template <typename T>
+void
+bucketListConsistent(
```
please find a better name for that function

MonsieurNicolas commented on src/bucket/BucketTests.cpp:
```diff
@@ -81,6 +81,57 @@ countEntries(std::shared_ptr<Bucket> bucket)
+checkBucketCount(BucketList& bl, uint32_t ledgerSeq, uint32_t level, bool isCurr)
+{
+    auto bucket = bl.getLevel(level).getCurr();
+    auto sizeOf = BucketList::sizeOfCurr;
```
rename `sizeOf` -> `sizeOfBucket`: it's too close to `sizeof`

jonjove replied:
Done

MonsieurNicolas commented on src/bucket/BucketTests.cpp:
```diff
@@ -81,6 +81,57 @@ countEntries(std::shared_ptr<Bucket> bucket)
 }
+
+void
+checkBucketCount(BucketList& bl, uint32_t ledgerSeq, uint32_t level, bool isCurr)
```
I don't think this function actually counts anything: it's checking if the bucket's lower bound matches its ledger entries.
Also, can't we also check the upper bound at the same time?

MonsieurNicolas commented on src/invariant/BucketListIsConsistentWithDatabase.cpp:
```diff
@@ -0,0 +1,134 @@
+    uint64_t nAccounts = 0, nTrustLines = 0, nOffers = 0, nData = 0;
+    for (Bucket::InputIterator iter(bucket); iter; ++iter)
+    {
+        auto const& e = *iter;
```
to make this code bullet proof, we should also check that 2 consecutive entries are strictly ordered

jonjove replied:
Done

vogel commented on src/bucket/BucketTests.cpp:
```diff
@@ -81,6 +81,68 @@ countEntries(std::shared_ptr<Bucket> bucket)
+// If pred is false for ledger < L and true for ledger >= L then
+// binarySearchForLedger will return L.
+uint32_t
+binarySearchForLedger(
```
strange that there is no std algorithm for that...

vogel commented on src/historywork/ApplyBucketsWork.cpp:
```diff
@@ -99,9 +85,18 @@ void
+    HistoryStateBucket const& i = mApplyState.currentBuckets.at(mLevel);
     if (mApplying || i.snap != binToHex(level.getSnap()->getHash()))
     {
+        if (!mApplying)
```
this will only be called when `i.snap != binToHex(level.getSnap()->getHash()` is true
what about moving that out of outside if so it becames
```
auto isCurrentSnapshot = i.snap != binToHex(level.getSnap()->getHash();
if (!isCurrentSnapshot) { ...}
if (mApplying || !isCurrentSnapshot) {...}
```

vogel commented on src/historywork/ApplyBucketsWork.cpp:
```diff
@@ -111,6 +106,15 @@ ApplyBucketsWork::onStart()
     {
+        if (!mApplying)
+        {
+            uint32_t oldestLedger =
```
this looks exactly like code above, maybe a method?

vogel commented on src/ledger/AccountFrame.cpp:
```diff
@@ -353,6 +356,47 @@ AccountFrame::countObjects(soci::session& sess)
 }
 
+uint64_t
+AccountFrame::countObjects(soci::session& sess,
```
I think that you could use LedgerRange class here (from catchup-at PR, now in master).
countObjects(sess, int, int) looks cryptic from API point of view
countObjectsInRange(sess, int, int) is better
but
countObjects(sess, LedgerRange) is best ;)

vogel commented on src/ledger/AccountFrame.cpp:
```diff
@@ -353,6 +356,47 @@ AccountFrame::countObjects(soci::session& sess)
+}
+
+void
+AccountFrame::deleteAccounts(Database& db, uint32_t oldestLedger)
```
By the name and argument types I would expect that to remove accounts on specific ledger, not from to.
I think it needs better name, like deleteAccountModifiedOnOrAfter (seems long) or deleteAccountFrom (not as clear).

Do you have any better idea?

MonsieurNicolas commented:
@jonjove it looks like this PR needs to be rebased again

jonjove commented:
Rebased, cherry-picked from #1364, and changed `Application::create` to `createTestApplication` in BucketListIsConsistentWithDatabaseTests.cpp, InvariantTests.cpp, and BucketTests.cpp

MonsieurNicolas commented:
r+ 811b01c67e8bf10e7fa2fd739a24d8b97a540cbc

MonsieurNicolas commented:
github sorts by date not by commit... trying again

MonsieurNicolas commented:
r+ 4df5454b75ff2326288d0d50c02e93852b408aa3

jonjove commented:
Fixed compile error

MonsieurNicolas commented:
r+ 35dbc2cae4b0d7ef9cd4325bd93073d5edfa7e1f

MonsieurNicolas commented:
r+ 5bedc6771bea9c146a3bbfa55e2dbeecedd263e4
